tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php

```diff
@@ -2,39 +2,94 @@
 
 namespace Doctrine\Tests\DBAL\Driver;
 
+use Doctrine\DBAL\Driver\IBMDB2\DB2Statement;
+use Doctrine\DBAL\Driver\Mysqli\MysqliStatement;
+use Doctrine\DBAL\Driver\OCI8\OCI8Statement;
+use Doctrine\DBAL\Driver\SQLAnywhere\SQLAnywhereStatement;
+use Doctrine\DBAL\Driver\SQLSrv\SQLSrvStatement;
 use Doctrine\DBAL\Driver\Statement;
 use Doctrine\DBAL\Driver\StatementIterator;
+use Doctrine\DBAL\Portability\Statement as PortabilityStatement;
+use IteratorAggregate;
+use PHPUnit\Framework\MockObject\MockObject;
+use Traversable;
+use function extension_loaded;
 
 class StatementIteratorTest extends \Doctrine\Tests\DbalTestCase
 {
-    public function testGettingIteratorDoesNotCallFetch()
+    /**
+     * @dataProvider statementProvider()
+     */
+    public function testGettingIteratorDoesNotCallFetch(string $class)
     {
-        $stmt = $this->createMock(Statement::class);
+        /** @var IteratorAggregate|MockObject $stmt */
+        $stmt = $this->createPartialMock($class, ['fetch', 'fetchAll', 'fetchColumn']);
         $stmt->expects($this->never())->method('fetch');
         $stmt->expects($this->never())->method('fetchAll');
         $stmt->expects($this->never())->method('fetchColumn');
 
+        $stmt->getIterator();
+    }
+
+    public function testIteratorIterationCallsFetchOncePerStep()
+    {
+        $stmt = $this->createMock(Statement::class);
+        $this->configureStatement($stmt, $calls);
+
         $stmtIterator = new StatementIterator($stmt);
-        $stmtIterator->getIterator();
+
+        $this->assertIterationCallsFetchOncePerStep($stmtIterator, $calls);
     }
 
-    public function testIterationCallsFetchOncePerStep()
+    /**
+     * @dataProvider statementProvider()
+     */
+    public function testStatementIterationCallsFetchOncePerStep(string $class)
+    {
+        $stmt = $this->createPartialMock($class, ['fetch']);
+        $this->configureStatement($stmt, $calls);
+        $this->assertIterationCallsFetchOncePerStep($stmt, $calls);
+    }
+
+    private function configureStatement(MockObject $stmt, &$calls) : void
```

Comment:
can haz int typehint?
I am pretty sure @Ocramius will hate the reference here. :laughing:

Reply:
We've been through that a couple of times. Sometimes they just make sense.